lib/marisa/grimoire/vector/vector.h

```diff
@@ -231,8 +235,19 @@ class Vector {
     MARISA_DEBUG_IF(new_buf.get() == NULL, MARISA_MEMORY_ERROR);
     T *new_objs = reinterpret_cast<T *>(new_buf.get());
 
-    for (std::size_t i = 0; i < size_; ++i) {
-      new (&new_objs[i]) T(objs_[i]);
+#if __cplusplus >= 201103L
+    constexpr bool trivially_copyable = std::is_trivially_copyable<T>::value;
+#else
+    const bool trivially_copyable = false;
+#endif
+    if (trivially_copyable) {
+      std::memcpy(reinterpret_cast<void*>(new_objs),
+                  reinterpret_cast<const void*>(objs_),
+                  sizeof(T) * size_);
+    } else {
+      for (std::size_t i = 0; i < size_; ++i) {
+        new (&new_objs[i]) T(objs_[i]);
+      }
     }
     for (std::size_t i = 0; i < size_; ++i) {
       objs_[i].~T();
```

Comment:
This loop can be moved inside the else clause above.

If a class is trivially copyable, it must have a trivial destructor.

From https://en.cppreference.com/w/cpp/named_req/TriviallyCopyable:

> **Requirements**
> ...
> * Trivial non-deleted destructor 

From https://en.cppreference.com/w/cpp/language/destructor#Trivial_destructor:

> Objects with trivial destructors don't require a delete-expression and may be disposed of by simply deallocating their storage.

Reply:
There are quite a few other `~T` calls, so it seems inconsistent to avoid it here, but not the other places.

The loop is optimized away with `-O1` or higher anyway: https://godbolt.org/z/-zH4gd

Reply:
I was a bit confused by the destructor calls in the trivially copyable case. It made me stop, question my understanding of C++, and look up the definition of trivially copyable.